### PR TITLE
Add non-darwin search paths

### PR DESCRIPTION
**Pull request by jmschonfeld**
This PR brings the swift-corelibs-foundation implementation of FileManager search paths for non-Darwin platforms to swift-foundation. Particularly, it brings functionality to lookup requested directories for Windows, and it adds functionality to check XDG environment variables / configuration files on platforms such as linux. The unit tests also validate this by checking the number of returned paths (to not rely on specific paths which can be modified by the system) and for XDG systems we can use the environment variables to validate specific paths.

This also tweaks the existing implementation to support more of `URL` now that the type is available.

### CODE REVIEW

**Comment by jmschonfeld** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/FileManager/SearchPaths/FileManager+WindowsSearchPaths.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if os(Windows)
+
+private func _url(for id: KNOWNFOLDERID) -> URL {
+    var pszPath: PWSTR?
+    let hResult: HRESULT = withUnsafePointer(to: id) { id in
+        SHGetKnownFolderPath(id, DWORD(KF_FLAG_DEFAULT.rawValue), nil, &pszPath)
```
Comment (compnerd) [inline Sources/FoundationEssentials/FileManager/SearchPaths/FileManager+WindowsSearchPaths.swift:18]: I think it makes sense to just put `KF_FLAG_DEFAULT` as an internal var in `WinSDK+Extensions.swift`

File: Sources/FoundationEssentials/FileManager/SearchPaths/FileManager+WindowsSearchPaths.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if os(Windows)
+
+private func _url(for id: KNOWNFOLDERID) -> URL {
+    var pszPath: PWSTR?
+    let hResult: HRESULT = withUnsafePointer(to: id) { id in
```
Comment (compnerd) [inline Sources/FoundationEssentials/FileManager/SearchPaths/FileManager+WindowsSearchPaths.swift:17]: `hr` is the traditional spelling for the variable.

File: Sources/FoundationEssentials/FileManager/SearchPaths/FileManager+WindowsSearchPaths.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        SHGetKnownFolderPath(id, DWORD(KF_FLAG_DEFAULT.rawValue), nil, &pszPath)
+    }
+    precondition(hResult >= 0, "SHGetKnownFolderpath failed \(GetLastError())")
```
Comment (compnerd) [inline Sources/FoundationEssentials/FileManager/SearchPaths/FileManager+WindowsSearchPaths.swift:20]: I think that `SUCCEEDED(hr)` would be a better check than `hr >= 0`.

**Comment by jmschonfeld** [comment]
@swift-ci please test

**Review by compnerd — APPROVED** [review]

**Review by parkera — APPROVED** [review]

**Comment by jmschonfeld** [comment]
@swift-ci please test

**Comment by jmschonfeld** [comment]
@swift-ci please test

**Pull request merged**